components/brave_rewards/core/common/request_signer.cc

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_rewards/core/common/request_signer.h"
+
+#include "base/base64.h"
+#include "base/strings/strcat.h"
+#include "crypto/sha2.h"
+#include "url/gurl.h"
+
+namespace brave_rewards::internal {
+
+namespace {
+
+const char* GetMethodString(mojom::UrlMethod method) {
+  switch (method) {
+    case mojom::UrlMethod::GET:
+      return "get";
+    case mojom::UrlMethod::PUT:
+      return "put";
+    case mojom::UrlMethod::POST:
+      return "post";
+    case mojom::UrlMethod::PATCH:
+      return "patch";
+    case mojom::UrlMethod::DEL:
+      return "delete";
+  }
+}
+
+std::string GetRequestTarget(const mojom::UrlRequest& request) {
+  GURL url(request.url);
+  if (!url.is_valid()) {
+    return "";
+  }
+  return base::StrCat({GetMethodString(request.method), " ", url.path()});
+}
+
+}  // namespace
+
+RequestSigner::RequestSigner(const std::string& key_id, Signer signer)
+    : key_id_(key_id), signer_(signer) {}
+
+RequestSigner::~RequestSigner() = default;
+
+RequestSigner::RequestSigner(const RequestSigner&) = default;
+RequestSigner& RequestSigner::operator=(const RequestSigner&) = default;
+
+std::string RequestSigner::GetDigest(base::span<const uint8_t> content) {
+  return "SHA-256=" + base::Base64Encode(crypto::SHA256Hash(content));
+}
+
+std::optional<RequestSigner> RequestSigner::FromRewardsWallet(
+    const mojom::RewardsWallet& rewards_wallet) {
+  auto signer = Signer::FromRecoverySeed(rewards_wallet.recovery_seed);
+  if (!signer) {
+    return std::nullopt;
+  }
+  return RequestSigner(rewards_wallet.payment_id, *signer);
+}
+
+bool RequestSigner::SignRequest(mojom::UrlRequest& request) {
+  std::string request_target = GetRequestTarget(request);
+  if (request_target.empty()) {
+    return false;
+  }
+
+  auto headers = GetSignedHeaders(request_target, request.content);
```

Comment:
Does request content contain any additional headers beyond digest and request-target or do we not sign over them?

Reply:
> I don't think any of our current endpoint implementations supply additional headers that would need to be signed
How did we determine this? It seems odd to me to only sign part of the request. I'd think it would be useful for us to sign the entire set of content and headers here instead of just selecting a subset. However, I'm not certain if we have any additional headers that will be going with this request.

Reply:
Yeah this appears like it will create a regression in [this call](https://github.com/brave/brave-core/pull/21467/files#r1445541437) where some of the headers that were previously signed now won't be.

Reply:
Nevermind, I made a mistake - the expected behavior is the same where we previously signed just a subset of headers and are still doing the same. https://github.com/brave/brave-core/pull/21467/files#diff-5fc0b225f5d25ee7f514478ed03449ca53059e18ec6da1905f17946d34a2dc4dL35-L65